hashing/hashing.cc

```diff
@@ -143,4 +118,74 @@ void Hashing::computeFileHashes(const vector<shared_ptr<core::File>> &files, spd
     }
 }
 
+vector<ast::ParsedFile> Hashing::indexAndComputeFileHashes(unique_ptr<core::GlobalState> &gs,
+                                                           const realmain::options::Options &opts,
+                                                           spdlog::logger &logger, vector<core::FileRef> &files,
+                                                           WorkerPool &workers,
+                                                           const unique_ptr<const OwnedKeyValueStore> &kvstore) {
+    auto asts = realmain::pipeline::index(gs, files, opts, workers, kvstore);
+    ENFORCE_NO_TIMER(asts.size() == files.size());
+
+    // In parallel, rewrite ASTs to an empty GlobalState and use them for hashing.
+    shared_ptr<ConcurrentBoundedQueue<size_t>> fileq = make_shared<ConcurrentBoundedQueue<size_t>>(asts.size());
+    for (size_t i = 0; i < asts.size(); i++) {
+        auto copy = i;
+        fileq->push(move(copy), 1);
+    }
+
+    logger.debug("Computing state hashes for {} files", asts.size());
+
+    const core::GlobalState &sharedGs = *gs;
+    shared_ptr<BlockingBoundedQueue<vector<pair<size_t, unique_ptr<const core::FileHash>>>>> resultq =
+        make_shared<BlockingBoundedQueue<vector<pair<size_t, unique_ptr<const core::FileHash>>>>>(asts.size());
+    Timer timeit(logger, "computeFileHashes");
+    workers.multiplexJob("lspStateHash", [fileq, resultq, &asts, &sharedGs, &logger]() {
+        unique_ptr<Timer> timeit;
+        vector<pair<size_t, unique_ptr<const core::FileHash>>> threadResult;
+        int processedByThread = 0;
+        size_t job;
+        {
+            for (auto result = fileq->try_pop(job); !result.done(); result = fileq->try_pop(job)) {
+                if (result.gotItem()) {
+                    if (timeit == nullptr) {
+                        timeit = make_unique<Timer>(logger, "computeFileHashesWorker");
```

Comment:
Avoids a race condition when outer function exits before a thread gets work.

Ugh, we really should introduce a better abstraction for parallelization.

Reply:
Yeah. I was thinking of defining an interface that contained the functionality used in most of these loops:

* Enqueues work
* Encodes the name of the task (used for metrics/timers)
* Filter function to ignore items that aren't important
* Function that performs the transformation and returns one object for that transformation
* Logic for interruption in LSP mode
* (Maybe) Special cases loops over small number of items to not use WorkerPool
  * We have this optimization in index; dunno how much it saves us.

I started poking at this before but got caught in the weeds trying to support _all_ of the things we do in these passes.

Reply:
There's an added complication of being thread safe, too. Some of that functionality is accessed from coordinator thread, others is accessed from worker thread.